### PR TITLE
Feature/263 blog listing FE

### PR DESCRIPTION
**Pull request by BrianXu20**
Expose the BE to the blog_listing.html page 
Ticket Link: https://projects.torchbox.com/projects/tbxcom/tickets/263

### CODE REVIEW

File: tbx/blog/models.py [hunk 1]
```diff
@@ -58,10 +66,10 @@ def serve(self, request):
         # Get blog_posts
         blog_posts = self.blog_posts
 
-        # Filter by tag
```
Comment (BrianXu20) [inline tbx/blog/models.py:61]: Make the blogs filter by services just like https://torchbox.com/blog/

File: tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.html [hunk 2]
```diff
@@ -5,19 +5,66 @@
...
+      if (this.readyState == 4 && this.status == 200) {
+      // if success
+        console.log(this.responseText);
+        }else if(this.readyState == 4){
+        // if error
+        page-=1;
+        }
+    }
+
+</script>
```
Comment (SharmaineLim) [inline tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.html:69]: Javascript goes under `tbx/static_src/javascript/components/insert-name-here.js`

You can copy the style of `cookie-message.js` for the structure to follow. It's also a good way to make the Javascript available across blog listing and work listing.

You can add something like `data-fetch-target="blog-pages"` to the "See more" button (you can make this a setting on see-more.html, like 
```
<a class="..."{% if link %} href="{{ link }}"{% elif fetch_target %} data-fetch-target="{{ fetch_target }}"{% endif %}
```

And then have this on the blog listing page:
```
<div class="blog-listing__list" data-fetch-destination="blog-pages">
```

We can talk more about it while you work on it if you'd like.

Also, please indent proper to make the JS easier to read.

File: tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.html [hunk 3]
```diff
@@ -5,19 +5,66 @@
...
-        <div class="blog-listing__content">
-
-            <div class="blog-listing__list">
-                {% for blog in page.blogs %}
-                    {% include "patterns/molecules/blog-item/blog-item.html" with item=blog %}
-                {% endfor %}
-            </div>
-            
-            {% include "patterns/atoms/see-more/see-more.html" with item=see_more %}
+{% include "patterns/molecules/title-block/title-block.html" with item=self tags=self.related_services %}
```
[the remark this answers is not in the document]
Comment (BrianXu20) [reply tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.html:8]: ok. I will update the context

File: tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.html (hunk 2 above)
Comment (BrianXu20) [reply tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.html:69]: okay

Comment (BrianXu20) [reply tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.html:69]: cookie-message.js looks like it's from Gatsby framework. Should I copy it's style ?

Comment (SharmaineLim) [reply tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.html:69]: I'm pretty sure it's more object-oriented than Gatsby, so yes, you can copy cookie-message.

```
class Abc {
    constructor(node) { ... }
    myMethod() { ... }
}
```

Gatsy framework is more on using JS to render HTML:
```
() => {
    return (
        <div>
            <p>This is html code.</p>
        </div>
    );
}
```

File: tbx/static_src/javascript/components/see-more-posts.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	{
+		this.node = node;
+		this.seeMoreButton = document.querySelector("[data-posts-fetch-target]");
+		this.target = this.node.dataset.postsDestination;
+		this.targetNode = document.querySelector(
+			`[data-posts-destination="${this.target}"]`
+		);
+
+		this.url = new URL(this.node.baseURI);
+		this.params = new URLSearchParams(this.url.search);
```
Comment (SharmaineLim) [inline tbx/static_src/javascript/components/see-more-posts.js:15]: Nitpick: It looks like we don't use the info from `this.params` in a different method/function, so we can also replace this with a local variable instead of adding it to `this`.

File: tbx/static_src/javascript/components/see-more-posts.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+		this.nextPage = 2;
+		this.filter = this.params.get("filter");
+
+		this.bindEvents();
+	}
+
+	loadMorePostsAjax()
+	{
+		//set parameters
+		this.params = new URLSearchParams();
```
Comment (SharmaineLim) [inline tbx/static_src/javascript/components/see-more-posts.js:24]: Nitpick: Since we don't use the info from `this.params` in a different method/function, so we can also replace this with a local variable instead of adding it to `this`.

File: tbx/static_src/javascript/components/see-more-posts.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+		{
+			method: "GET",
+			headers: new Headers(
+			{
+				"X-Requested-With": "XMLHttpRequest",
+			}),
+		});
+
+		fetch(request)
+			.then(function (response)
```
Comment (SharmaineLim) [inline tbx/static_src/javascript/components/see-more-posts.js]: You can use lambda functions for these for consistency:

```
then(response => { ... })
```
or
```
then((response) => { ... })
```

File: tbx/static_src/javascript/components/see-more-posts.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+				else
+				{
+					// handle empty response
+					console.log("empty response")
+				}
+			}.bind(this))
+			.catch(function (error)
+			{
+				// If there is any error you will catch them here
+				console.log(error);
```
Comment (SharmaineLim) [inline tbx/static_src/javascript/components/see-more-posts.js:58]: Let's avoid using `console.log()`. I think the default behaviour is that errors get output to the console anyway.

File: tbx/static_src/javascript/components/see-more-posts.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+					// parse to html
+					return response.text();
+				}
+				throw new Error(response.status + ": " + response.statusText);
+			})
+			.then(function (html)
+			{
+				if (html.trim().length > 0)
+				{
+					this.targetNode.lastElementChild.insertAdjacentHTML('afterend', html);
```
Comment (SharmaineLim) [inline tbx/static_src/javascript/components/see-more-posts.js:49]: We can shorten this to:
```
this.targetNode.appendChild(html);
```

File: tbx/static_src/javascript/components/see-more-posts.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+					this.nextPage += 1;
+				}
+				else
+				{
+					// handle empty response
+					console.log("empty response")
```
Comment (SharmaineLim) [inline tbx/static_src/javascript/components/see-more-posts.js]: In this case, I think we'd want to hide the "See more" button.

e.g. add a class `see-more--hidden` and add CSS to make it happen

(Also, let's not leave `console.log`s on production)

File: tbx/static_src/javascript/components/see-more-posts.js [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+			{
+				// If there is any error you will catch them here
+				console.log(error);
+			});
+
+	}
+
+	bindEvents()
+	{
+		if (!this.seeMoreButton)
```
Comment (SharmaineLim) [inline tbx/static_src/javascript/components/see-more-posts.js]: Would this ever be true?

File: tbx/static_src/javascript/components/see-more-posts.js (hunk 8 above)
Comment (BrianXu20) [reply tbx/static_src/javascript/components/see-more-posts.js:49]: appendChild() method accepts an Object, so I can’t just pass raw html into that method.  I guess it will be good just leave it as it is ?

File: tbx/static_src/javascript/components/see-more-posts.js (hunk 10 above)
Comment (BrianXu20) [reply tbx/static_src/javascript/components/see-more-posts.js]: it will be always false, I will just remove this part

File: tbx/static_src/sass/components/_see-more.scss [hunk 11]
```diff
@@ -19,4 +19,8 @@
             padding: 15px 30px;
         }
     }
+
+    &__display__none{
```
Comment (SharmaineLim) [inline tbx/static_src/sass/components/_see-more.scss:23]: For CSS, we follow the BEM, which is basically:

```
.component
.component--modifier
.component__subcomponent
.component__subcomponent--modifier
```

in HTML:
```
<div class="container container--wide">
```

So for this one, we should probably name it
```
&--hidden {
```

File: tbx/static_src/sass/components/_see-more.scss [hunk 12]
```diff
@@ -19,4 +19,8 @@
...
+        display:none;
```
Comment (SharmaineLim) [inline tbx/static_src/sass/components/_see-more.scss:24]: Nitpick: Please add a space between `display:` and `none;`

File: tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.yaml [hunk 13]
```diff
@@ -3,13 +3,14 @@ context:
...
+      - name: Digital products
+        slug: digital-products
+      - name: Digital products
+        slug: digital-products
+      - name: Digital products
+        slug: digital-products
+      - name: Digital products
+        slug: digital-products
   blog_posts:
     - {
```
[the remark this answers is not in the document]
Comment (BrianXu20) [reply tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.yaml:15]: blog_posts and related_services are working, except I am having a hard time on setting  blog_posts.has_next

[the remark this answers is not in the document]
Comment (SharmaineLim) [reply tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.yaml]: Try adding it as a tag:

```
tags:
  if:
    blog_posts.has_next:
      raw: True
```

[the remark this answers is not in the document]
Comment (SharmaineLim) [reply tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.yaml]: Also you don't need to remove the curlies if it was working

[the remark this answers is not in the document]
Comment (BrianXu20) [reply tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.yaml]: I was thinking to make the yaml file less lines of code, it's working also on this way

File: tbx/project_styleguide/templates/patterns/pages/blog/includes/blog_listing.yaml [hunk 14]
```diff
@@ -0,0 +1,32 @@
+context:
+  blog_posts:
+    - title: We already dropped Drupal - maybe it’s time for you to do the same.,
```
Comment (SharmaineLim) [inline tbx/project_styleguide/templates/patterns/pages/blog/includes/blog_listing.yaml:3]: Nitpick: Commas at the end not needed anymore (same for the original yaml file)

File: tbx/static_src/javascript/components/see-more-posts.js [hunk 15]
```diff
@@ -39,7 +39,7 @@ class SeeMorePosts {
                     // parse to html
                     return response.text();
                 }
-                throw new Error(response.status + ": " + response.statusText);
+                throw new Error(`${response.status}: ${response.statusText}`);
```
Comment (SharmaineLim) [inline tbx/static_src/javascript/components/see-more-posts.js:42]: :+1:

File: tbx/static_src/javascript/components/see-more-posts.js [hunk 16]
```diff
@@ -9,16 +9,16 @@ class SeeMorePosts {
         this.targetNode = document.querySelector(
             `[data-fetch-destination="${node.dataset.fetchTarget}"]`
         );
-        this.MAX_PAGES = parseInt(this.targetNode.getAttribute('data-fetch-max-pages'));
+        this.MAX_PAGES = parseInt(this.targetNode.getAttribute('data-fetch-max-pages'),10);
```
Comment (SharmaineLim) [inline tbx/static_src/javascript/components/see-more-posts.js:12]: Nitpick: Please add a space after the comma. Later on we'll be adding linter checks and this will get flagged.

**Review by SharmaineLim — APPROVED** [review]
Pretty much good to go, I think.

File: tbx/blog/models.py [hunk 17]
```diff
@@ -58,31 +66,39 @@ def serve(self, request):
...
+            # use page to filter
+            page = request.GET.get('page')
+            try:
+                blog_posts = paginator.page(page)
+            except PageNotAnInteger:
+                blog_posts = paginator.page(1)
+            except EmptyPage:
+                blog_posts = None
+
+            return render(request, 'patterns/pages/blog/includes/blog_listing.html', {
```
Comment (SharmaineLim) [inline tbx/blog/models.py:88]: Almost missed this. This doesn't follow the pattern library structure we have. Might be better to put this is under `patterns/organisms/blog-listing/blog-listing.html`

File: tbx/blog/models.py [hunk 18]
```diff
@@ -44,6 +47,11 @@ def get_popular_tags(self):
         # Getting them individually to preserve the order
         return [Tag.objects.get(id=tag['tag']) for tag in popular_tags[:10]]
 
+    @property
+    def related_services(self):
+        services = Service.objects.all()
+        return services
```
Comment (SharmaineLim) [inline tbx/blog/models.py:53]: I'm bothered by this. `related_services` isn't actually a property of `BlogIndexPage`, so this be better to just include in `serve`

**Review by SharmaineLim — APPROVED** [review]
Happy with this now.

File: tbx/project_styleguide/templates/patterns/organisms/blog-listing/blog-listing.yaml [hunk 19]
```diff
@@ -0,0 +1,32 @@
+context:
+  blog_posts:
+    - title: We already dropped Drupal - maybe it’s time for you to do the same.
+      date: 20 Oct '20
+      author_name: Paul Vetch
+      author_role: Strategy Director
```
Comment (SharmaineLim) [inline tbx/project_styleguide/templates/patterns/organisms/blog-listing/blog-listing.yaml:6]: I forgot, could you update these please?

File: tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.yaml [hunk 20]
```diff
@@ -2,50 +2,48 @@ context:
...
+      slug: digital-products
+
+    - name: Digital products
+      slug: digital-products
+
+  blog_posts:
+    - title: We already dropped Drupal - maybe it’s time for you to do the same.
+      date: 20 Oct '20
+      author_name: Paul Vetch
+      author_role: Strategy Director
```
Comment (SharmaineLim) [inline tbx/project_styleguide/templates/patterns/pages/blog/blog_listing.yaml:22]: These too
